packages/node-core/src/indexer/mmrMigrate.service.ts

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import {existsSync} from 'fs';
+import {DEFAULT_WORD_SIZE} from '@subql/common';
+import {NodeConfig, getExistingProjectSchema, getLogger, PgBasedMMRDB} from '@subql/node-core';
+import {FileBasedDb} from '@subql/x-merkle-mountain-range';
+import {Logging, QueryTypes, Sequelize} from 'sequelize';
+
+const logger = getLogger('mmr-migrate');
+
+export enum MMRMigrateErrorCode {
+  SchemaCreationError = 'SCHEMA_CREATION_ERROR',
+  MMRFileNotFoundError = 'MMR_FILE_NOT_FOUND_ERROR',
+}
+
+export interface MMRMigrateErrorContext {
+  [key: string]: any;
+}
+
+export class MMRMigrateError extends Error {
+  readonly code: MMRMigrateErrorCode;
+  readonly context?: MMRMigrateErrorContext;
+
+  constructor(message: string, code: MMRMigrateErrorCode, context?: MMRMigrateErrorContext) {
+    super(message);
+    this.name = 'MMRMigrateError';
+    this.code = code;
+    this.context = context;
+  }
+
+  toString(): string {
+    let errorString = `${this.name}: ${this.message}\nCode: ${this.code}`;
+    if (this.context) {
+      errorString += `\nContext:\n${JSON.stringify(this.context, null, 2)}`;
+    }
+    return errorString;
+  }
+}
+
+export enum MigrationDirection {
+  FileToDb = 'fileToDb',
+  DbToFile = 'dbToFile',
+}
+
+export class MMRMigrateService {
+  constructor(private nodeConfig: NodeConfig, private sequelize: Sequelize) {}
+
+  async migrate(direction: MigrationDirection): Promise<void> {
+    try {
+      if (direction === MigrationDirection.FileToDb && !existsSync(this.nodeConfig.mmrPath)) {
+        throw new MMRMigrateError(
+          `MMR file not found: ${this.nodeConfig.mmrPath}`,
+          MMRMigrateErrorCode.MMRFileNotFoundError
+        );
+      } else if (direction === MigrationDirection.DbToFile) {
+        const schema = await getExistingProjectSchema(this.nodeConfig, this.sequelize);
+        if (!schema) {
+          throw new MMRMigrateError(`Schema for MMR DB does not exist`, MMRMigrateErrorCode.SchemaCreationError);
+        }
+        const tables = await this.sequelize.query('SELECT tablename FROM pg_tables WHERE schemaname = ?', {
+          replacements: [schema],
+          type: QueryTypes.SELECT,
+        });
+        const tableNames = tables.map((table: any) => table.tablename);
+
+        if (!tableNames.includes('_mmrs')) {
+          throw new MMRMigrateError(
+            'The _mmr table does not exist in the database',
```

Comment:
Users don't need to know the internal workings.
```suggestion
            'MMR data is not found in the database, unable to migrate.',
```